train.py

```diff
@@ -239,6 +247,10 @@ def loss_fn(pred, labels):
         f"({gpu_mem_stats.max_reserved_pct:.2f}%)"
     )
 
+    # TODO(whc) Make a wrapper class for Optimizers, Schedulers, expose 'step' API for convenience
```

Comment:
I wonder what is feeling around using an `nn.ModuleList` instead of a normal Python `list` so that we can still treat it as a single root module.

Reply:
feels morally wrong to me. the modules are not really connected sequentially, and calling the .forward of the modulelist would be semantically invalid.

Reply:
otoh, it is a nice hack in many ways. it would let us wrap one optimizer, easily pass to DCP, (well, scratch that: it'd change fqn's on us wouldn't it?)

Reply:
`nn.ModuleList` does not implement `forward` (unlike `nn.Sequential`), so I think semantically it is just a container.

Reply:
so what would the proposed flow be?

```
modules = ModuleList(model_parts)

apply_spmd(_) # would we use 'modules' here and deal with applying fsdp one more time to the root, but operating over the first level of children for the current tp/dp logic?  Or would we apply this to model_parts and then manually apply fsdp one more time outside to 'modules'?

checkpoint.save(model_parts)  # i think we can't use the outer wrapper during checkpoint save/load right or have to adjust fqn manually there?
```

Reply:
I think on the FSDP side, the main constraint is actually just to not call `fully_shard(ModuleList)` since then FSDP forward hooks would never run (we actually error today with FSDP2 if you try to call `fully_shard(ModuleList)`). If you were to use a `ModuleList`, then we should just apply `fully_shard` to the children of the `ModuleList`.

I am not as familiar with the checkpointing constraints. If using a `ModuleList` causes too much checkpointing headache, then feel free to not use a `ModuleList`. Either way works for FSDP.